client/extensions/woocommerce/state/data-layer/orders/index.js

```diff
@@ -54,82 +54,93 @@ const onDeleteSuccess = action => dispatch => {
-	],
-	[ WOOCOMMERCE_ORDERS_REQUEST ]: [ dispatchRequest( requestOrders, receivedOrders, apiError ) ],
-};
-
-export function apiError( { dispatch }, action, error ) {
+export function apiError( action, error ) {
 	debug( 'API Error: ', error );
 
 	if ( action.failureAction ) {
-		dispatch( { ...action.failureAction, error } );
+		return { ...action.failureAction, error };
 	}
 }
 
-export function requestOrders( { dispatch }, action ) {
+export function requestOrders( action ) {
 	const { siteId, query } = action;
 	const queryString = stringify( omitBy( query, val => '' === val ) );
 	action.failureAction = failOrders( siteId, query );
 
-	dispatch( request( siteId, action ).getWithHeaders( `orders?${ queryString }` ) );
+	return request( siteId, action ).getWithHeaders( `orders?${ queryString }` );
 }
 
-export function receivedOrders( { dispatch }, action, { data } ) {
+export function receivedOrders( action, { data } ) {
 	const { siteId, query } = action;
 	const { body, headers } = data;
 	const total = Number( headers[ 'X-WP-Total' ] );
 
-	dispatch( updateOrders( siteId, query, body, total ) );
+	return updateOrders( siteId, query, body, total );
 }
 
-export function requestOrder( { dispatch }, action ) {
+export function requestOrder( action ) {
 	const { siteId, orderId } = action;
 	action.failureAction = failOrder( siteId, orderId );
 
-	dispatch( request( siteId, action ).get( `orders/${ orderId }` ) );
+	return request( siteId, action ).get( `orders/${ orderId }` );
 }
 
-export function receivedOrder( { dispatch }, action, { data } ) {
+export function receivedOrder( action, { data } ) {
 	const { siteId, orderId } = action;
-	dispatch( updateOrder( siteId, orderId, data ) );
+	return updateOrder( siteId, orderId, data );
 }
 
-export function sendOrder( { dispatch }, action ) {
+export function sendOrder( action ) {
 	const { siteId, orderId, order } = action;
 	if ( isFinite( orderId ) ) {
-		dispatch( request( siteId, action ).post( `orders/${ orderId }`, order ) );
-	} else {
-		dispatch( request( siteId, action ).post( 'orders', order ) );
+		return request( siteId, action ).post( `orders/${ orderId }`, order );
 	}
+	return request( siteId, action ).post( 'orders', order );
 }
 
-export function onOrderSaveSuccess( { dispatch }, action, { data } ) {
-	const { siteId, orderId } = action;
-	// Make sure we have a success function, and a new order ID
-	if ( 'function' === typeof action.onSuccess && 'undefined' !== typeof data.id ) {
-		action.onSuccess( dispatch, data.id );
-	}
-	dispatch( fetchCounts( siteId ) );
-	dispatch( saveOrderSuccess( siteId, orderId, data ) );
+export function onOrderSaveSuccess( action, { data } ) {
+	return dispatch => {
+		const { siteId, orderId } = action;
+		// Make sure we have a success function, and a new order ID
+		if ( 'function' === typeof action.onSuccess && 'undefined' !== typeof data.id ) {
+			action.onSuccess( dispatch, data.id );
```

Comment:
This `onSuccess` method ([added here](https://github.com/Automattic/wp-calypso/blob/master/client/extensions/woocommerce/app/order/header.js#L90-L115)) expects `dispatch` to be present to show a success notice which is why I used a thunk here.